### PR TITLE
New: Grid- infra for Gap control [ED-9760]

### PR DESCRIPTION
**Pull request by davseve**

### CODE REVIEW

File: assets/dev/scss/editor/panel/controls/_gap.scss [hunk 1]
```diff
@@ -0,0 +1,104 @@
+.elementor-control-type-gap {
```
Comment (louiswol94) [inline assets/dev/scss/editor/panel/controls/_gap.scss:1]: This entire CSS file looks like a copy of `_dimensions.scss`. Maybe we can extract the common code to `_proportions_common.scss` and then make `_gap.scss` and `_dimensions.scss` individually much smaller and have them import the common file. What do you think?

File: includes/controls/gap.php [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+			'column' => esc_html__( 'Column', 'elementor' ),
+		];
+		?>
+		<div class="elementor-control-field">
+			<label class="elementor-control-title">{{{ data.label }}}</label>
+			<?php $this->print_units_template(); ?>
+			<div class="elementor-control-input-wrapper">
+				<ul class="elementor-control-gaps">
+					<?php
+					foreach ( $dimensions as $dimension_key => $dimension_title ) :
```
Comment (louiswol94) [inline includes/controls/gap.php:95]:
```suggestion
					foreach ( $this->get_dimensions() as $dimension_key => $dimension_title ) :
```
Maybe just a bit better for readability? You can decide.

File: includes/controls/gap.php [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+									echo $dimension_key; // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
+									?> ) ) {
+										print( placeholder.<?php
+										// PHPCS - the variable $dimension_key is a plain text.
+										echo $dimension_key; // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
+										?> );
+									}
+								} else {
+								print( placeholder );
+								} #>"
```
Comment (louiswol94) [inline includes/controls/gap.php:115]: There are nested if statements here. Can this be improved like this for example?

```
const placeholder = view.getControlPlaceholder();
if ( _.isObject( placeholder ) && !_.isUndefined( placeholder[dimension_key] ) ) {
  print( placeholder[dimension_key] );
} else {
  print( placeholder );
}
```

**Review by louiswol94 — CHANGES_REQUESTED** [review]

File: assets/dev/js/editor/controls/gap.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+		return ui;
+	},
+
+	getPossibleDimensions() {
+		return [
+			'row',
+			'column',
+		];
+	},
+} );
```
Comment (louiswol94) [inline assets/dev/js/editor/controls/gap.js:21]: Add blank lines above and below `module.exports = ControlGapItemView;`

File: assets/dev/js/editor/controls/gap.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
+let ControlDimensionsView = require( 'elementor-controls/dimensions' ),
```
Comment (louiswol94) [inline assets/dev/js/editor/controls/gap.js:1]: Check JSLint warnings

**Review by louiswol94 — CHANGES_REQUESTED** [review]

File: assets/dev/js/editor/controls/gap.js (hunk 4 above)
Comment (davseve) [reply assets/dev/js/editor/controls/gap.js:21]: Done

File: assets/dev/scss/editor/panel/controls/_gap.scss (hunk 1 above)
Comment (davseve) [reply assets/dev/scss/editor/panel/controls/_gap.scss:3]: Hi, What I ended up doing was to inherit the `dimensions.scss` into the `gap.scss`.
I thought of creating a common base, the issue is that this is scss, and the main wrapper class is different, therefor AFAIU, the base should have the control's entire code. and then in each of the individual files I need to override the control-related parts.
Please correct me if I am wrong

File: includes/controls/gap.php (hunk 3 above)
Comment (louiswol94) [reply includes/controls/gap.php]: I don't see this change :-)

File: includes/controls/gap.php [hunk 6]
```diff
@@ -0,0 +1,151 @@
+<?php
```
Comment (louiswol94) [inline includes/controls/gap.php:1]: As discussed, check if you can extend Dimensions here?

**Review by louiswol94 — CHANGES_REQUESTED** [review]

File: assets/dev/scss/editor/panel/controls/_dimensions.scss [hunk 7]
```diff
@@ -1,102 +1 @@
...
-			color: var(--e-a-color-txt-accent);
-			border-color: var(--e-a-border-color-bold);
-
-			.elementor-unlinked {
-				display: none;
-			}
-		}
-	}
-}
+@include dimensions( 'dimension' );
```
Comment (louiswol94) [inline assets/dev/scss/editor/panel/controls/_dimensions.scss:1]: @daviedR Where did you create this `@mixin`?

**Review by louiswol94 — CHANGES_REQUESTED** [review]

**Review by louiswol94 — DISMISSED** [review]

File: includes/controls/gaps.php [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+/**
+ * Elementor gap control.
+ *
+ * A base control for creating gap control. Displays input fields for two values,
+ * row/column, hight/width/ and the option to link them together.
```
Comment (hein-obox) [inline includes/controls/gaps.php:12]: Suggested: 'A base control for creating a gap control. Displays input fields for two values,
 row/column, height/width and the option to link them together'

File: includes/controls/dimensions.php [hunk 9]
```diff
@@ -84,55 +97,48 @@ protected function get_default_settings() {
...
-								placeholder = view.getControlPlaceholder();
-								if ( _.isObject( placeholder ) ) {
-									if ( ! _.isUndefined( placeholder.<?php
+									placeholder = view.getControlPlaceholder();
+									if ( _.isObject( placeholder ) && ! _.isUndefined( placeholder.<?php
 										// PHPCS - the variable $dimension_key is a plain text.
 										echo $dimension_key; // phpcs:ignore WordPress.Security.EscapeOutput.OutputNotEscaped
 									?> ) ) {
-										print( placeholder.<?php
+											print( placeholder.<?php
```
Comment (hein-obox) [inline includes/controls/dimensions.php:120]: Indentation.

File: includes/controls/gaps.php (hunk 8 above)
Comment (louiswol94) [reply includes/controls/gaps.php]: @hein-obox Done

File: includes/controls/dimensions.php (hunk 9 above)
Comment (louiswol94) [reply includes/controls/dimensions.php:120]: @hein-obox This was a bit of a refactor, so it just looks like the code moved unnecessarily in git.

**Review by hein-obox — DISMISSED** [review]

File: includes/controls/dimensions.php [hunk 10]
```diff
@@ -73,6 +77,15 @@ protected function get_default_settings() {
 		);
 	}
 
+	protected function dimensions() {
```
[the remark this answers is not in the document]
Comment (louiswol94) [reply includes/controls/dimensions.php]: @TzviRabinovitch Done

**Review by TzviRabinovitch — APPROVED** [review]

**Pull request merged**